File: maker/golang/xdb/searcher.go
// Copyright 2022 The Ip2Region Authors. All rights reserved.
// Use of this source code is governed by a Apache2.0-style
// license that can be found in the LICENSE file.

// ---
// Ip2Region database v2.0 searcher.
// this is part of the maker for testing and validate.
// please use the searcher in binding/golang for production use.
// And this is a Not thread safe implementation.

package xdb

import (
	"encoding/binary"
	"fmt"
	"os"
)

type Searcher struct {
	version *Version

	handle *os.File

	// header info
	header []byte

	// use it only when this feature enabled.
	// Preload the vector index will reduce the number of IO operations
	// thus speedup the search process
	vectorIndex []byte
}

func NewSearcher(version *Version, dbFile string) (*Searcher, error) {
	handle, err := os.OpenFile(dbFile, os.O_RDONLY, 0600)
	if err != nil {
		return nil, err
	}

	return &Searcher{
		version: version,

		handle: handle,
		header: nil,

		vectorIndex: nil,
	}, nil
}

func (s *Searcher) Close() {
	if s.handle != nil {
		err := s.handle.Close()
		if err != nil {
			return
		}
	}
}

// LoadVectorIndex load and cache the vector index for search speedup.
// this will take up VectorIndexRows x VectorIndexCols x VectorIndexSize bytes memory.
func (s *Searcher) LoadVectorIndex() error {
	// loaded already
	if s.vectorIndex != nil {
		return nil
	}

	// load all the vector index block
	_, err := s.handle.Seek(HeaderInfoLength, 0)
	if err != nil {
		return fmt.Errorf("seek to vector index: %w", err)
	}

	var buff = make([]byte, VectorIndexLength)
	rLen, err := s.handle.Read(buff)
	if err != nil {
		return err
	}

	if rLen != len(buff) {
		return fmt.Errorf("incomplete read: readed bytes should be %d", len(buff))
	}

	s.vectorIndex = buff
	return nil
}

// ClearVectorIndex clear preloaded vector index cache
func (s *Searcher) ClearVectorIndex() {
	s.vectorIndex = nil
}

// Search find the region for the specified ip address
func (s *Searcher) Search(ip []byte) (string, int, error) {
	// version check
	if len(ip) != s.version.Bytes {
		return "", 0, fmt.Errorf("invalid ip address(%s expected)", s.version.Name)
	}

	// locate the segment index block based on the vector index
	var ioCount = 0
	var il0, il1, bytes, tBytes = int(ip[0]), int(ip[1]), len(ip), len(ip) << 1
	var idx = il0*VectorIndexCols*VectorIndexSize + il1*VectorIndexSize
	var sPtr, ePtr = uint32(0), uint32(0)
	if s.vectorIndex != nil {
		sPtr = binary.LittleEndian.Uint32(s.vectorIndex[idx:])
		ePtr = binary.LittleEndian.Uint32(s.vectorIndex[idx+4:])
	} else {
		pos, err := s.handle.Seek(int64(HeaderInfoLength+idx), 0)
		if err != nil {
			return "", ioCount, fmt.Errorf("seek to vector index %d: %w", HeaderInfoLength+idx, err)
		}

		ioCount++
		var buff = make([]byte, VectorIndexSize)
		rLen, err := s.handle.Read(buff)
		if err != nil {
			return "", ioCount, fmt.Errorf("read vector index at %d: %w", pos, err)
		}

		if rLen != len(buff) {
			return "", ioCount, fmt.Errorf("incomplete read: readed bytes should be %d", len(buff))
		}

		sPtr = binary.LittleEndian.Uint32(buff)
		ePtr = binary.LittleEndian.Uint32(buff[4:])
	}

	//log.Printf("vIndex=%s", vIndex)
	// binary search the segment index to get the region
	var segIndexSize = uint32(s.version.SegmentIndexSize)
	var dataLen, dataPtr = 0, uint32(0)
	var buff = make([]byte, segIndexSize)
	var l, h = 0, int((ePtr - sPtr) / segIndexSize)
	for l <= h {
		// log.Printf("l=%d, h=%d", l, h)
		m := (l + h) >> 1
		p := sPtr + uint32(m)*segIndexSize
		// log.Printf("m=%d, p=%d", m, p)
		_, err := s.handle.Seek(int64(p), 0)
		if err != nil {
			return "", ioCount, fmt.Errorf("seek to segment block at %d: %w", p, err)
		}

		ioCount++
		rLen, err := s.handle.Read(buff)
		if err != nil {
			return "", ioCount, fmt.Errorf("read segment index at %d: %w", p, err)
		}

		if rLen != len(buff) {
			return "", ioCount, fmt.Errorf("incomplete read: readed bytes should be %d", len(buff))
		}

		// decode the data step by step to reduce the unnecessary calculations
		if s.version.IPCompare(ip, buff[0:bytes]) < 0 {
			h = m - 1
		} else if s.version.IPCompare(ip, buff[bytes:tBytes]) > 0 {
			l = m + 1
		} else {
			dataLen = int(binary.LittleEndian.Uint16(buff[tBytes:]))
			dataPtr = binary.LittleEndian.Uint32(buff[tBytes+2:])
			break
		}
	}

	if dataLen == 0 {
		return "", ioCount, nil
	}

	// load and return the region data
	_, err := s.handle.Seek(int64(dataPtr), 0)
	if err != nil {
		return "", ioCount, fmt.Errorf("seek to data block at %d: %w", dataPtr, err)
	}

	ioCount++
	var regionBuff = make([]byte, dataLen)
	rLen, err := s.handle.Read(regionBuff)
	if err != nil {
		return "", ioCount, fmt.Errorf("read region data at %d: %w", dataPtr, err)
	}

	if rLen != dataLen {
		return "", ioCount, fmt.Errorf("incomplete read: readed bytes should be %d", dataLen)
	}

	return string(regionBuff), ioCount, nil
}

// LoadXdbHeader load the header info from the specified handle
func LoadXdbHeader(handle *os.File) ([]byte, error) {
	_, err := handle.Seek(0, 0)
	if err != nil {
		return nil, fmt.Errorf("seek to the header: %w", err)
	}

	var buff = make([]byte, HeaderInfoLength)
	rLen, err := handle.Read(buff)
	if err != nil {
		return nil, err
	}

	if rLen != len(buff) {
		return nil, fmt.Errorf("incomplete read: readed bytes should be %d", len(buff))
	}

	return buff, nil
}

// LoadXdbHeaderFromFile load header info from the specified db file path
func LoadXdbHeaderFromFile(dbFile string) ([]byte, error) {
	handle, err := os.OpenFile(dbFile, os.O_RDONLY, 0600)
	if err != nil {
		return nil, fmt.Errorf("open xdb file `%s`: %w", dbFile, err)
	}

	defer func(handle *os.File) {
		_ = handle.Close()
	}(handle)

	header, err := LoadXdbHeader(handle)
	if err != nil {
		return nil, err
	}

	return header, nil
}

File: binding/csharp/IP2Region/xdb/Searcher.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace IP2Region.xdb
{
    public class Searcher
    {

        // constant defined copied from the xdb maker
        public static int HeaderInfoLength = 256;
        public static int VectorIndexRows = 256;
        public static int VectorIndexCols = 256;
        public static int VectorIndexSize = 8;
        public static int SegmentIndexSize = 14;

        // random access file handle for file based search
        private readonly Stream handle;

        private int ioCount = 0;

        public int IOCount => this.ioCount;


        // vector index.
        // use the byte[] instead of VectorIndex entry array to keep
        // the minimal memory allocation.
        private readonly byte[] vectorIndex;

        // xdb content buffer, used for in-memory search
        private readonly byte[] contentBuff;


        // --- static method to create searchers


        public static Searcher NewWithFileOnly(String dbPath)
        {
            return new Searcher(dbPath, null, null);
        }

        public static Searcher NewWithVectorIndex(String dbPath, byte[] vectorIndex)
        {
            return new Searcher(dbPath, vectorIndex, null);
        }

        public static Searcher NewWithBuffer(byte[] cBuff)
        {
            return new Searcher(null, null, cBuff);
        }

        // --- End of creator
        public Searcher(string dbFile, byte[] vectorIndex, byte[] cBuff)
        {
            if (cBuff != null)
            {
                this.handle = null;
                this.vectorIndex = null;
                this.contentBuff = cBuff;
            }
            else
            {
                this.handle = File.OpenRead(dbFile);
                this.vectorIndex = vectorIndex;
                this.contentBuff = null;
            }
        }

        public void Close()
        {
            if (this.handle != null) this.handle.Close();
        }

        public string Search(string ipStr)
        {
            var ip = checkIP(ipStr);
            return Search(ip);
        }

        public string Search(long ip)
        {
            // reset the global counter
            this.ioCount = 0;
            // locate the segment index block based on the vector index
            int sPtr = 0, ePtr = 0;
            int il0 = (int)((ip >> 24) & 0xFF);
            int il1 = (int)((ip >> 16) & 0xFF);
            int idx = il0 * VectorIndexCols * VectorIndexSize + il1 * VectorIndexSize;
            if (vectorIndex != null)
            {
                sPtr = GetInt(vectorIndex, idx);
                ePtr = GetInt(vectorIndex, idx + 4);
            }
            else if (contentBuff != null)
            {
                sPtr = GetInt(contentBuff, HeaderInfoLength + idx);
                ePtr = GetInt(contentBuff, HeaderInfoLength + idx + 4);
            }
            else
            {
                byte[] vectorBuff = new byte[VectorIndexSize];
                Read(HeaderInfoLength + idx, vectorBuff);
                sPtr = GetInt(vectorBuff, 0);
                ePtr = GetInt(vectorBuff, 4);
            }

            // binary search the segment index block to get the region info
            byte[] buff = new byte[SegmentIndexSize];
            int dataLen = -1, dataPtr = -1;
            int l = 0, h = (ePtr - sPtr) / SegmentIndexSize;
            while (l <= h)
            {
                int m = (l + h) >> 1;
                int p = sPtr + m * SegmentIndexSize;

                // read the segment index
                Read(p, buff);
                long sip = GetIntLong(buff, 0);
                if (ip < sip)
                {
                    h = m - 1;
                }
                else
                {
                    long eip = GetIntLong(buff, 4);
                    if (ip > eip)
                    {
                        l = m + 1;
                    }
                    else
                    {
                        dataLen = GetInt2(buff, 8);
                        dataPtr = GetInt(buff, 10);
                        break;
                    }
                }
            }

            // empty match interception
            // System.out.printf("dataLen: %d, dataPtr: %d\n", dataLen, dataPtr);
            if (dataPtr < 0) return null;

            // load and return the region data
            byte[] regionBuff = new byte[dataLen];
            Read(dataPtr, regionBuff);
            //return new String(regionBuff, "utf-8");
            return Encoding.UTF8.GetString(regionBuff);
        }

        protected virtual void Read(int offset, byte[] buffer)
        {
            // check the in-memory buffer first
            if (contentBuff != null)
            {
                // @TODO: reduce data copying, directly decode the data ?
                //System.arraycopy(contentBuff, offset, buffer, 0, buffer.length);
                Array.Copy(contentBuff, offset, buffer, 0, buffer.Length);
                return;
            }

            // read from the file handle
            if (handle == null) throw new ArgumentNullException(nameof(handle));
            handle.Seek(offset, SeekOrigin.Begin);

            this.ioCount++;
            var rLen = handle.Read(buffer, 0, buffer.Length);
            if (rLen != buffer.Length) throw new IOException("incomplete read: read bytes should be " + buffer.Length);
        }

        // --- static cache util function
        public static Header LoadHeader(Stream stream)
        {
            stream.Seek(0, SeekOrigin.Begin);
            var buffer = new byte[HeaderInfoLength];
            stream.Read(buffer, 0, HeaderInfoLength);
            return new Header(buffer);
        }
        public static Header LoadHeaderFromFile(string dbPath)
        {
            using (var fs = File.OpenRead(dbPath)) return LoadHeader(fs);
        }
        public static byte[] LoadVectorIndex(Stream stream)
        {
            stream.Seek(HeaderInfoLength, SeekOrigin.Begin);
            int len = VectorIndexRows * VectorIndexCols * SegmentIndexSize;
            var buff = new byte[len];
            var rLen = stream.Read(buff, 0, buff.Length);
            if (rLen != len) throw new IOException("incomplete read: read bytes should be " + len);
            return buff;
        }
        public static byte[] LoadVectorIndexFromFile(string dbPath)
        {
            using (var fs = File.OpenRead(dbPath)) return LoadVectorIndex(fs);
        }
        public static byte[] LoadContent(Stream stream)
        {
            stream.Seek(0, SeekOrigin.Begin);
            using (var ms = new MemoryStream())
            {
                stream.CopyTo(ms);
                return ms.ToArray();
            }
        }

        public static byte[] LoadContentFromFile(string dbPath)
        {
            using (var fs = File.OpenRead(dbPath)) return LoadContent(fs);
        }

        public static int GetInt2(byte[] b, int offset)
        {
            return (
                (b[offset++] & 0x000000FF) |
                (b[offset] & 0x0000FF00)
            );
        }
        public static int GetInt(byte[] b, int offset)
        {
            return (
                ((b[offset++]) & 0x000000FF) |
                ((b[offset++] << 8) & 0x0000FF00) |
                ((b[offset++] << 16) & 0x00FF0000) |
                (int)((b[offset] << 24) & 0xFF000000)
            );
        }
        public static long GetIntLong(byte[] b, int offset)
        {
            return (
                ((b[offset++] & 0x000000FFL)) |
                ((b[offset++] << 8) & 0x0000FF00L) |
                ((b[offset++] << 16) & 0x00FF0000L) |
                ((b[offset] << 24) & 0xFF000000L)
            );
        }

        /* long int to ip string */
        public static string Long2ip(long ip)
        {
            return string.Join(".", (ip >> 24) & 0xFF, (ip >> 16) & 0xFF, (ip >> 8) & 0xFF, (ip) & 0xFF);
        }

        public static byte[] shiftIndex = { 24, 16, 8, 0 };

        /* check the specified ip address */
        public static long checkIP(String ip)
        {
            String[]
            ps = ip.Split('.');
            if (ps.Length != 4) throw new Exception("invalid ip address `" + ip + "`");

            long ipDst = 0;
            for (int i = 0; i < ps.Length; i++)
            {
                int val = Convert.ToInt32(ps[i]);
                if (val > 255)
                {
                    throw new Exception("ip part `" + ps[i] + "` should be less then 256");
                }
                ipDst |= ((long)val << shiftIndex[i]);
            }

            return ipDst & 0xFFFFFFFFL;
        }
    }
}
